network_simulator/host.py

```diff
@@ -23,4 +27,65 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: host that receives packet
+        :ivar packet: packet received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host and \
+        packet.packet_id != ROUTER_PACKET_DEFAULT_ID
+
+        self.host = host
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        self.flow_packets_received[self.packet.flow_id] = self.packet.packet_id
+        self.flow_packets_received.sort()
+        
+        statistics.host_packet_received(self.host, self.packet,
+                                        main_event_loop.global_clock_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        If packet was Ack, look up Flow and say packet was received.
+        Schedule FlowReceivedAckEvent.
+        Populate AckPacket for selective repeat.
+
+        If packet was data packet, generate Ack and schedule via
+        DeviceToLink event unless packet already received earlier.
+
+        Ackpackets, set original data packet's start time for RTT.
+        AckPacket ID same as DataPacket ID
+        :param main_event_loop: schedule new Events
+        :param statistics: update statistics
+        """
+
+        if isinstance(self.packet, AckPacket):
+            main_event_loop.schedule_new_event_with_delay(self, self.flow,
+                                                          self.packet)
+            ack_flow_id = self.packet.flow_id
+            for i in self.flow_packets_received:
+                if self.flow_packets_received[i].flow_id == ack_flow_id:
+                    self.flow_packets_received[i].packet_id = self.packet_id
+        if isinstance(self.packet, DataPacket) and \
```

Comment:
For this part, you need to check if this packet has been received before by this Host. This actually needs to be in run() though. Here's what you do:

1) Add an ivar called "packet_previously_received" to __init__ (update the ivar documentation too). This is a boolean that will be True if the Packet had previously been received by the Host.

2) In run(), _before_ you bisect.insort() the new Packet ID into the list of Packets received by the Flow, you need to check that it's not already there. Here is a function that can help you do this (this should go somewhere in the HostReceivedPacketEvent class):
```
def elem_in_list(elem, my_list):
    # ind is where this element *would* be inserted into the list in order to keep
    # it in sorted order.
    ind = bisect.bisect_left(my_list, elem)
    if ind != len(my_list) and my_list[ind] == elem:
        return True
    else:
        return False
```

You can test out this function in a terminal to see that it does in fact work. What this is doing is simple. First, it uses binary search (the "bisect" call) to look up where we would insert "elem" into "my_list" in order to keep it in sorted order. This index is what I called "ind". If the element at "ind" is in fact elem, then elem is in the list. Otherwise, elem cannot be in the list. This code is faster than searching the list element by element since it uses binary search, not linear search.

What you want to do is, in run(), let self.packet_previously_received = elem_in_list(packet.packet_id, old_packets_received). Do this after the line where you define "old_packets_received" (see above comment). If the packet was previously received, then you should not add it to the old_packets_received list via insort.

3) Here, in schedule_new_events(), check if instance(self.packet, DataPacket) and not self.packet_previously_received. That is the condition you want to satisfy before sending ACKs. I will comment on how to generate ACKs below.